### ISSUE TITLE
`Image.network()` in `AnimatedList` crash when not connected to the Internet

### ISSUE THREAD

**Issue by mregnauld** [flutter/flutter#129106]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues) (found #20910 but it doesn't solve my problem)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Go to: https://github.com/mregnauld/flutter-image-network-crash-demo
2. Disconnect your device from the Internet
3. Launch the app
4. Scroll the list, maybe for quite a few seconds

### Expected results

I should be able to scroll without any error in the logs and without seeing the error page.
Every `Image.network()` that can not download the image should just display the error widget returned by `errorBuilder`.

### Actual results

At some point, you should see a new page displaying the error (that you can also see in the logs).

### Code sample

<details><summary>Code sample</summary>

```dart
import 'dart:async';
import 'dart:developer';
import 'dart:ui';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:test_image_list/error.dart';

GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>();

void main()
{
  runZonedGuarded(() async
  {
    // lancement :
    WidgetsFlutterBinding.ensureInitialized();

    // debug et tests :
    debugPrintMarkNeedsPaintStacks = false;
    debugProfilePaintsEnabled = false;
    debugProfileBuildsEnabled = false;
    debugRepaintRainbowEnabled = false;
    debugRepaintTextRainbowEnabled = false;
    debugPaintLayerBordersEnabled = false;

    // meilleur rendu des dégradés :
    Paint.enableDithering = true;

    // gestion erreurs :
    PlatformDispatcher.instance.onError = (error, stack) {
      log("$error\n\n$stack");
      return true;
    };

    FlutterError.onError = (FlutterErrorDetails details) {
      FlutterError.dumpErrorToConsole(details);
      log(details.toString());
      WidgetsBinding.instance.addPostFrameCallback((_) {
        Navigator.push(
          _navigatorKey.currentState!.context,
          MaterialPageRoute(builder: (context) => ErrorPage(error: details.toString())),
        );
      });
    };
    runApp(const MyApp());
  }, (Object error, StackTrace stack) {
    log("$error\n\n$stack");
    Navigator.push(
      _navigatorKey.currentState!.context,
      MaterialPageRoute(builder: (context) => ErrorPage(error: "$error\n\n$stack")),
    );
  });
}

class MyApp extends StatelessWidget
{
  const MyApp({super.key});

  @override
  Widget build(BuildContext context)
  {
    return MaterialApp(
      navigatorKey: _navigatorKey,
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget
{
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage>
{

  final _scrollController = ScrollController();
  
  
  @override
  Widget build(BuildContext context)
  {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: const Text("List demo"),
      ),
      body: SafeArea(
        child: AnimatedList(
          controller: _scrollController,
          initialItemCount: 100,
          itemBuilder: (context, index, animation)
          {
            return Container(
              constraints: const BoxConstraints(minHeight: 64),
              padding: const EdgeInsets.all(8),
              child: Row(
                mainAxisSize: MainAxisSize.max,
                mainAxisAlignment: MainAxisAlignment.start,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  Image.network(
                    "http://s519716619.onlinehome.fr/test/image.jpg?id=$index",
                    width: 80,
                    height: 60,
                    cacheWidth: (MediaQuery.of(context).devicePixelRatio * 80).round(),
                    cacheHeight: (MediaQuery.of(context).devicePixelRatio * 60).round(),
                    errorBuilder: (context, error, stackTrace) {
                      return Container(
                        width: 80,
                        height: 60,
                        color: Colors.red,
                      );
                    },
                  ),
                  const SizedBox(width: 8),
                  Expanded(
                    child: Text(
                      "Item $index",
                    ),
                  ),
                ]
              ),
            );
          },
        ),
      ),
    );
  }
  
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
I/flutter (22671): ══╡ EXCEPTION CAUGHT BY IMAGE RESOURCE SERVICE ╞════════════════════════════════════════════════════
I/flutter (22671): The following SocketException was thrown resolving an image codec:
I/flutter (22671): Failed host lookup: 's519716619.onlinehome.fr' (OS Error: No address associated with hostname, errno
I/flutter (22671): = 7)
I/flutter (22671): 
I/flutter (22671): When the exception was thrown, this was the stack:
I/flutter (22671): #0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:719:35)
I/flutter (22671): #1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1877:26)
I/flutter (22671): #2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
I/flutter (22671): #3      Socket._startConnect (dart:io-patch/socket_patch.dart:2100:22)
I/flutter (22671): #4      Socket.startConnect (dart:io/socket.dart:766:21)
I/flutter (22671): #5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2486:20)
I/flutter (22671): #6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2924:12)
I/flutter (22671): #7      _HttpClient._getConnection (dart:_http/http_impl.dart:2929:12)
I/flutter (22671): #8      _HttpClient._openUrl (dart:_http/http_impl.dart:2784:12)
I/flutter (22671): #9      _HttpClient.getUrl (dart:_http/http_impl.dart:2627:48)
I/flutter (22671): #10     NetworkImage._loadAsync (package:flutter/src/painting/_network_image_io.dart:124:59)
I/flutter (22671): #11     NetworkImage.loadImage (package:flutter/src/painting/_network_image_io.dart:84:14)
I/flutter (22671): #12     ResizeImage.loadImage (package:flutter/src/painting/image_provider.dart:1436:58)
I/flutter (22671): #13     ImageProvider.resolveStreamForKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:529:39)
I/flutter (22671): #14     ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:382:22)
I/flutter (22671): #15     ImageProvider.resolveStreamForKey (package:flutter/src/painting/image_provider.dart:526:81)
I/flutter (22671): #16     ScrollAwareImageProvider.resolveStreamForKey (package:flutter/src/widgets/scroll_aware_image_provider.dart:105:19)
I/flutter (22671): #17     ScrollAwareImageProvider.resolveStreamForKey.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/scroll_aware_image_provider.dart:99:33)
I/flutter (22671): (elided 10 frames from dart:async)
I/flutter (22671): 
I/flutter (22671): Image provider: NetworkImage("http://s519716619.onlinehome.fr/test/image.jpg?id=44", scale: 1.0)
I/flutter (22671): Image key: NetworkImage("http://s519716619.onlinehome.fr/test/image.jpg?id=44", scale: 1.0)
I/flutter (22671): ════════════════════════════════════════════════════════════════════════════════════════════════════
I/flutter (22671): Another exception was thrown: SocketException: Failed host lookup: 's519716619.onlinehome.fr' (OS Error: No address associated with hostname, errno = 7)
[log] ══╡ EXCEPTION CAUGHT BY IMAGE RESOURCE SERVICE ╞═════════════════
The following SocketException was thrown resolving an image
codec:
Failed host lookup: 's519716619.onlinehome.fr' (OS Error: No
address associated with hostname, errno = 7)

When the exception was thrown, this was the stack:
#0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:719:35)
#1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1877:26)
#2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
#3      Socket._startConnect (dart:io-patch/socket_patch.dart:2100:22)
#4      Socket.startConnect (dart:io/socket.dart:766:21)
#5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2486:20)
#6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2924:12)
#7      _HttpClient._getConnection (dart:_http/http_impl.dart:2929:12)
#8      _HttpClient._openUrl (dart:_http/http_impl.dart:2784:12)
#9      _HttpClient.getUrl (dart:_http/http_impl.dart:2627:48)
#10     NetworkImage._loadAsync (package:flutter/src/painting/_network_image_io.dart:124:59)
#11     NetworkImage.loadImage (package:flutter/src/painting/_network_image_io.dart:84:14)
#12     ResizeImage.loadImage (package:flutter/src/painting/image_provider.dart:1436:58)
#13     ImageProvider.resolveStreamForKey.<anonymous closure> (package:flutter/src/painting/image_provider.dart:529:39)
#14     ImageCache.putIfAbsent (package:flutter/src/painting/image_cache.dart:382:22)
#15     ImageProvider.resolveStreamForKey (package:flutter/src/painting/image_provider.dart:526:81)
#16     ScrollAwareImageProvider.resolveStreamForKey (package:flutter/src/widgets/scroll_aware_image_provider.dart:105:19)
#17     ScrollAwareImageProvider.resolveStreamForKey.<anonymous closure>.<anonymous closure> (package:flutter/src/widgets/scroll_aware_image_provider.dart:99:33)
(elided 10 frames from dart:async)

Image provider:
  NetworkImage("http://s519716619.onlinehome.fr/test/image.jpg?id=44",
  scale: 1.0)
Image key:
  NetworkImage("http://s519716619.onlinehome.fr/test/image.jpg?id=44",
  scale: 1.0)
═════════════════════════════════════════════════════════════════

```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.4, on macOS 13.2 22D49 darwin-arm64, locale fr-FR)
    • Flutter version 3.10.4 on channel stable at /Users/mregnauld/FlutterSDK/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 682aa387cf (13 days ago), 2023-06-05 18:04:56 -0500
    • Engine revision 2a3401c9bb
    • Dart version 3.0.3
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc1)
    • Android SDK at /Users/mregnauld/AndroidSDK
    • Platform android-33, build-tools 33.0.0-rc1
    • ANDROID_HOME = /Users/mregnauld/AndroidSDK
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.77.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.60.0

[✓] Connected device (2 available)
    • SM M515F (mobile) • RF8NC0PSMMV • android-arm64 • Android 12 (API 31)
    • macOS (desktop)   • macos       • darwin-arm64  • macOS 13.2 22D49 darwin-arm64

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

**Comment by mregnauld** [comment]
@dam-ease Thanks for the suggestions.

I've already tried the second one but it doesn't work properly. For example, if I turn the screen off then turn it on, every image displays the error widget, even though it displayed the right image just before.
I also tried `cached_network_image` but I have the same problem.

My only workaround here is to ignore the `SocketException` in my `main()` function.

That being said:
- if I reconnect my device to the Internet, many (not all) images won't reload, even after scrolling up and down multiple times
- it shouldn't throw an error in the first place when I use `errorBuilder()` on the `Image.network()`
